### ISSUE TITLE
Coming up with possible network features to classify on

### ISSUE THREAD

**Issue by SandyaS72** [openconnectome/Statistical-Connectomics-Class#239]
What's a good way to start off coming up with network features that you want to classify on (degree-based, centrality-based, path length-based, etc... there's lots)? Do you just pick a handful that seem reasonable and try it, or is there a better way to figure out what kinds of features will work best?

**Comment by kristinmg** [comment]
I was classifying graphs based on networks features. To get some idea of what could work I picked a few features I wanted to observe and used bootstrapping to find the ones (or actually the nodes for each feature) that gave the most significant difference between the groups.

**Comment by ajulian3** [comment]
I agree with the above post. You could even use bootstrapping to relate qualitative binary information and eventually run a cross validation to determine detection accuracy. If you could be able to classify this information and assign it to subsets that would be further optimal. What are further ways to classify the network of connectomics data?

**Comment by sgomezr** [comment]
There's a section in the paper below that details some methods to compare variations between graphs and maybe classify it (it cites Joshua's paper XD). Anyways, yes @SandyaS72, the authors mention some research that focus on comparing graphs based on graph invariants (degree-based, centrality-based, path length-based, etc.)... so if these are used for comparison, they can also be used for classification?

R Cameron Craddock, Rosalia L Tungaraza and Michael P Milham; Connectomics and new approaches for analyzing human brain functional connectivity; GigaScience 2015.
http://www.biomedcentral.com/content/pdf/s13742-015-0045-x.pdf

**Comment by mrjiaruiwang** [comment]
You could generate "questions" instead of features and use a combination of decision trees for classification. An example question would be, for each edge, do the endpoints have a certain property like does it have more than one edge connected to it, or two, or multiple. You don't necessarily need features. It just happens that the classical methods of statistical learning like SVMs and neural networks use features.

**Comment by itsthefedora** [comment]
Feature selection is more of an art than a science, sadly, because higher-order relationships between features have a strong impact on how well they are for use in classification. Generally, the scheme is this: you have a gajillion different features that you could throw into your classification method of choice, and no idea *a priori* as to which ones will help and which ones won't. One entirely reasonable thing you could do is just rank the features by their R^2 correlation with the class label, and skim some small proportion of features off the top (or use significance statistics for correlation plus some multiple comparison testing---whatever makes you feel warm and fuzzy inside). Say you got 100 features out of this. The problem now is that those features could all be correlated *with each other*: that is, while the first feature really helped you identify the class, the remaining 99 just told you exactly the same thing as the first, and so adding in those features won't improve your classification accuracy.

Ignoring these concerns, a good first-order approach is to use mutual information with the class label, *i.e.*, quantize the features, rank them by empirical mutual information with the class label in the training dataset, and skim the "top" (however you want to define "top"). This has the upshot of giving you a sense of *how* informative features are: if you're doing binary classification, then you need 1 bit of conditional information out of your features to discern the class perfectly. This runs into the same issue as above, namely degradation of *true* conditional information content because of mutual information within the features.

A good higher-order technique to use is the decision tree: that is, you find a tree of binary "questions" you can ask about individual features that gives you the maximum amount of information about the class in your training set, then choose the features that were involved in those questions. This method, however, incurs a problem: finding globally optimal decision trees is hard (in both the loose and formal sense, though stepwise entropy reduction is a "decent" local method). One way around this is to use *random forests* instead: that is, you build a bunch of decision trees out of random subspaces of possible questions, and then select features that appear in some fraction of those trees' questions. Essentially, the idea is to "boost" your locally optimal feature selection: even though you really can only build decision trees in a greedy way, by aggregating a swath of randomly-made greedy trees, you can get an approximation to a more globally-optimal tree. These methods are generally used as classifiers themselves, but they provide a really good method of probing higher-order information dependencies, which makes them valuable for model-free feature selection.

**Comment by akim1** [comment]
If you can align your nodes in a consistent way, one seemingly infinite yet kind of silly features is just to check the presence of all the different combination of edges.

**Comment by adjordan** [comment]
Yeah, I just picked a few and checked which gave the most interesting results.
